docs/writing-controllers.md

```diff
@@ -0,0 +1,1447 @@
+  hub: EventEmitter;
+
+  constructor(/* ... */) {
+    // ...
+
+    this.hub = new EventEmitter();
+  }
+
+  doSomething() {
+    this.hub.emit('someEvent');
+  }
+}
+
+// === Client repo ===
+
+const fooController = new FooController();
+fooController.hub.on('someEvent', () => {
+  // respond to the event somehow
+});
+```
+
+However, this pattern can be replaced with the use of the messenger:
+
+✅ **The controller emits the `FooController:someEvent` event**
+
+```typescript
+/* === This repo: packages/foo-controller/src/FooController.ts === */
+
+const name = 'FooController';
+
+type FooControllerMessenger = RestrictedControllerMessenger<
+  typeof name,
+  never,
+  never,
+  never,
+  never
+>;
+
+class FooController extends BaseController<
+  'FooController',
+  // ...,
+  FooControllerMessenger
+> {
+  constructor({ messenger /*, ... */ }, { messenger: FooControllerMessenger }) {
+    super({ messenger /*, ... */ });
+  }
+
+  doSomething() {
+    this.messagingSystem.publish('FooController:someEvent');
+  }
+}
+
+// === Client repo ===
+
+const rootMessenger = new ControllerMessenger<
+  'FooController:someEvent',
+  never
+>();
+const fooControllerMessenger = rootMessenger.getRestricted({
+  name: 'FooController',
+});
+const fooController = new FooController({
+  messenger: fooControllerMessenger,
+});
+rootMessenger.subscribe('FooController:someEvent', () => {
+  // do something with the event
+});
+```
+
+## Define the `*:getState` action using the `ControllerGetStateAction` utility type
+
+Each controller needs a type for its `*:getState` action. The `ControllerGetStateAction` utility type from the `@metamask/base-controller` package should be used to define this type.
+
+The name of this type should be `${ControllerName}GetStateAction`.
+
+```typescript
+import type { ControllerGetStateAction } from '@metamask/base-controller';
+
+type FooControllerGetStateAction = ControllerGetStateAction<
+  'FooController',
+  FooControllerState
+>;
+```
+
+## Define the `*:stateChange` event using the `ControllerStateChangeEvent` utility type
+
+Each controller needs a type for its `*:stateChange` event. The `ControllerStateChangeEvent` utility type from the `@metamask/base-controller` package should be used to define this type.
+
+The name of this type should be `${ControllerName}StateChangeEvent`.
+
+```typescript
+import type { ControllerStateChangeEvent } from '@metamask/base-controller';
+
+type FooControllerStateChangeEvent = ControllerStateChangeEvent<
+  'FooController',
+  FooControllerState
+>;
+```
+
+## Use standard naming for action identifiers
+
+An action is equivalent to a function, and its identifier should be named in a similar fashion.
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerShowNotificationAction = {
+  type: `${typeof name}:showNotification`;
+  handler: () => void;
+};
+```
+
+## Use standard naming for action types
+
+Types for messenger actions should follow this naming:
```

Comment:
nit: "naming scheme" seems like the clearer wording.

```suggestion
## Use standard naming scheme for action types

Types for messenger actions should follow this naming scheme:
```

Reply:
Makes sense.